File: src/main/java/io/github/deepanshut041/sawdhyay/controller/CourseController.java
package io.github.deepanshut041.sawdhyay.controller;

import io.github.deepanshut041.sawdhyay.models.Course;
import io.github.deepanshut041.sawdhyay.models.Enrollment;
import io.github.deepanshut041.sawdhyay.services.CourseService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.servlet.ModelAndView;

import java.util.List;

@Controller
@RequestMapping("/courses")
public class CourseController {

    @Autowired
    private CourseService courseService;

    @RequestMapping(value = {"","/"}, method = RequestMethod.GET)
    public ModelAndView coursesPage(Model model){
        ModelAndView modelAndView = new ModelAndView();
        List<Course> courses = courseService.findAllCourses();
        modelAndView.addObject("courses", courses);
        modelAndView.setViewName("courses");
        return modelAndView;
    }

    @RequestMapping(value="/{id}", method = RequestMethod.GET)
    public ModelAndView getCourse(@PathVariable int id, Model model){
        ModelAndView modelAndView = new ModelAndView();
        Enrollment enrollment = new Enrollment();
        modelAndView.addObject("course", this.courseService.getCourseById(id));
        modelAndView.addObject("enrollments", enrollment);
        modelAndView.setViewName("course");
        return modelAndView;
    }

}

File: backend/src/main/kotlin/io/github/deepanshut041/swadhyay/web/controller/home/CourseController.kt
package io.github.deepanshut041.swadhyay.web.controller.home

import io.github.deepanshut041.swadhyay.data.entity.CourseEntity
import io.github.deepanshut041.swadhyay.services.CourseService
import io.github.deepanshut041.swadhyay.web.controller.home.CourseController.Companion.HOME_COURSE_URI
import io.swagger.v3.oas.annotations.tags.Tag
import kotlinx.coroutines.flow.produceIn
import kotlinx.coroutines.reactor.flux
import kotlinx.coroutines.reactor.mono
import org.springframework.data.repository.query.Param
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Flux

@RestController
@RequestMapping(HOME_COURSE_URI, produces = ["application/json"])
@Tag(name = "Home Course API", description = "This contains routes regarding course for public")
class CourseController(
        private val courseService: CourseService
) {

    @GetMapping
    fun getCourses(
            @RequestParam("level", required = false) level: String?,
            @RequestParam("category", required = false) category: String?
    ): Flux<CourseEntity> = flux {
        courseService.getAllCourses(level, category).produceIn(this)
    }

    @GetMapping("/{id}")
    fun getCourse(@PathVariable("id") id: String) = mono {
        courseService.getCourse(id)
    }

    companion object {
        const val HOME_COURSE_URI = "/api/home/courses"
    }
}